index.js

```diff
@@ -169,6 +272,198 @@ class Pawn extends Piece {
   get icon() {
     return this.isWhite() ? "♙" : "♟";
   }
+  onMove(move) {
+    this.hasMoved = true;
+  }
+}
+
+class Rook extends Piece {
+  constructor(game, board, color, row, file) {
+    super(game, board, color);
+    this.row = row;
+    this.file = file;
+    this.hasMoved = false;
+  }
+
+  name = "Rook";
+
+  get icon() {
+    return this.isWhite() ? "♖" : "♜";
+  }
+
+  get moves() {
+    const available = [];
+
+    this.verticalMove(this.row, this.file, available, "up");
+    this.verticalMove(this.row, this.file, available, "down");
+    this.horizontalMove(this.row, this.file, available, "right");
+    this.horizontalMove(this.row, this.file, available, "left");
+    return available;
+  }
+  onMove(move) {
+    this.hasMoved = true;
+  }
+}
+
+class Bishop extends Piece {
+  constructor(game, board, color, row, file) {
+    super(game, board, color);
+    this.row = row;
+    this.file = file;
+  }
+
+  name = "Bishop";
+
+  get moves() {
+    const available = [];
+
+    this.moveDiagonally(this.row, this.file, available, "up", "right");
+    this.moveDiagonally(this.row, this.file, available, "down", "left");
+    this.moveDiagonally(this.row, this.file, available, "up", "left");
+    this.moveDiagonally(this.row, this.file, available, "down", "right");
+    return available;
+  }
+
+  onMove(move) {
+    console.log(move);
+  }
+  get icon() {
+    return this.isWhite() ? "♗" : "♝";
+  }
+}
+
+class Knight extends Piece {
+  constructor(game, board, color, row, file) {
+    super(game, board, color);
+    this.row = row;
+    this.file = file;
+  }
+
+  name = "Knight";
+
+  elMove(row, file, arr, vert, horiz) {
+    const elSquare = [row + vert, file + horiz];
+
+    if (this.isValidSquare(...elSquare)) {
+      if (!this.isSquareOccupied(...elSquare)) {
+        arr.push(Move.fromSquare(elSquare, this));
+      } else if (
+        this.isSquareOccupied(...elSquare) &&
+        this.getSquareContent(...elSquare).color !== this.color
+      ) {
+        arr.push(
+          Move.fromSquare(elSquare, this, this.getSquareContent(...elSquare))
+        );
+      }
+    }
+    return arr;
+  }
+
+  get moves() {
+    const available = [];
+
+    this.elMove(this.row, this.file, available, 2, 1);
+    this.elMove(this.row, this.file, available, 1, 2);
+
+    this.elMove(this.row, this.file, available, -2, 1);
+    this.elMove(this.row, this.file, available, -1, 2);
+
+    this.elMove(this.row, this.file, available, -2, -1);
+    this.elMove(this.row, this.file, available, -1, -2);
+
+    this.elMove(this.row, this.file, available, 2, -1);
+    this.elMove(this.row, this.file, available, 1, -2);
+
+    return available;
```

Comment:
good job on this. Let's reuse our new favorite function
```suggestion
  get moves() {
    const directions = [[ 2, 1],[ 1, 2],[ -2, 1],[ -1, 2],[ -2, -1],[ -1, -2],[ 2, -1],[ 1, -2]]
    const available = directions.flatMap(dir => getLegalDirectionalMoves(this, direction, 1))
    return available
  }
```